File: biddo_client/lib/core/models/review.dart
import 'account.dart';

class Review {
  String id;
  int stars;

  String? description;
  String? auctionId;

  Account? reviewer;
  Account? reviewed;

  DateTime createdAt;
  DateTime? updatedAt;

  Review({
    required this.id,
    required this.stars,
    this.auctionId,
    this.description,
    this.reviewer,
    this.reviewed,
    required this.createdAt,
    this.updatedAt,
  });

  static Review fromJSON(dynamic data) {
    return Review(
      id: data['id'],
      stars: data['stars'],
      auctionId: data['auctionId'] ?? '',
      description: data['description'] ?? '',
      reviewer:
          data['reviewer'] != null ? Account.fromJSON(data['reviewer']) : null,
      reviewed:
          data['reviewed'] != null ? Account.fromJSON(data['reviewed']) : null,
      createdAt: DateTime.parse(data['createdAt']),
      updatedAt: DateTime.parse(data['updatedAt']),
    );
  }
}

File: core/domain/review.ts
import { Account } from './account'

export class Review {
  id: string
  stars: number

  description?: string
  auctionId?: string

  reviewer?: Account
  reviewed?: Account

  createdAt: Date
  updatedAt: Date

  constructor(params: Record<string, unknown> = {}) {
    this.id = params.id as string
    this.stars = params.stars as number
    this.description = params.description as string
    this.auctionId = params.auctionId as string
    this.reviewer = params.reviewer as Account
    this.reviewed = params.reviewed as Account
    this.createdAt = params.createdAt as Date
    this.updatedAt = params.updatedAt as Date
  }

  static fromJSON(data: Record<string, unknown>) {
    return new Review({
      id: data.id,
      stars: data.stars,
      description: data.description,
      auctionId: data.auctionId,
      reviewer:
        data.reviewer !== null ? Account.fromJSON(data.reviewer as Record<string, unknown>) : null,
      reviewed:
        data.reviewed !== null ? Account.fromJSON(data.reviewed as Record<string, unknown>) : null,
      createdAt: data.createdAt ? new Date(data.createdAt.toString()) : null,
      updatedAt: data.updatedAt ? new Date(data.updatedAt.toString()) : null,
    })
  }
}
